doc/lspconfig.txt

```diff
@@ -1,242 +1,659 @@
+    Returns either a filepath (string) or nil. The language server will only
+    start if the function returns a filepath.
 
-    ```bash
-    npm i -g pyright
-    ```
-2. Add the language server setup to your init.vim. The server name must match those found in the table of contents in [CONFIG.md](CONFIG.md)
+    If a root directory (string) is returned which is unique from any
+    previously returned root_dir, a new server will be spawned with that
+    root directory. See |lspconfig-root-directory| for more details
 
-    ```lua
-    lua << EOF
-    require'lspconfig'.pyright.setup{}
-    EOF
-    ```
-3. Create a project, this project must contain a file matching the root directory trigger. See [Automatically launching language servers](#Automatically-launching-language-servers) for additional info.
+- {name}
 
-    ```bash
-    mkdir test_python_project
-    cd test_python_project
-    git init
-    touch main.py
-    ```
-4. Launch neovim, the language server will now be attached and providing diagnostics (see `:LspInfo`)
+    `string`
+
+    Defaults to the server's name (`clangd`, `pyright`, etc.).
+
+- {filetypes}
+
+    `list[string] | nil`
+
+    Set of filetypes on which to attempt to resolve {root_dir}.
+
+    May be empty or server may specify a default value.
+
+- {autostart}
+
+    `bool` (default: true)
+
+    Controls if the `FileType` autocommand that launches a language server is
+    enabled. If `false`, allows for deferring language servers until manually
+    launched with `:LspStart` (|lspconfig-commands|).
+
+- {on_new_config}
+
+    `function(new_config, new_root_dir)`
+
+    Function executed after a root directory is detected. This is used to
+    modify the server configuration (including `cmd` itself). Most commonly,
+    this is used to inject additional arguments into `cmd`.
+
+    If overriding `on_new_config`, ensure that you read the
+    `on_new_config` defined in the source file of the default configuration
+    in lspconfig. The original `on_new_config` for a given server should
+    be likely be a subset of your new override
+
+Note: all entries passed to `setup{}` override the entry in the default
+configuration. There is no composition.
 
-    ```
-    nvim main.py
-    ```
-5. See [Keybindings and completion](#Keybindings-and-completion) for mapping useful functions and enabling omnifunc completion
 ==============================================================================
-5. Automatically launching language servers*lspconfig-automatically-launching-language-servers*
+OVERRIDES VIA SETUP                                  *lspconfig-setup-overrides*
 
+The previous keys allow for customizing a server without redefining the
+config. The following retains all default settings for gopls, but overrides
+the root pattern to only match `.git` directories:
+>
+  local lspconfig = require'lspconfig'
+  lspconfig.gopls.setup{
+    root_dir = lspconfig.util.root_pattern('.git');
+  }
+<
```

Comment:
```suggestion
config. The following retains all default settings for tsserver, but overrides
the filetypes to only allow it attach to `typesecript` files, instead of default preset
>
  local lspconfig = require'lspconfig'
  lspconfig.tsserver.setup{
    filetypes = {"typescript"}
  }
<
```

I think it's better to give an easier example here, given that you haven't explained `root_dir` or `root_pattern` yet.

Reply:
root_dir is what people will most common override, and I do briefly explain it.